HP-149: Button, Carousel, Select, Modal 컴포넌트 구현

minseung-gang opened the pull request:
## PR 타입

- [x] 기능 추가
- [ ] 버그 수정
- [x] 리팩토링
- [ ] 문서 수정
- [ ] 기타

## 반영 브랜치

HP-149/ms →  dev
 

## 작업 사항
- 공통 Button 컴포넌트 / StyledButton 컴포넌트 구현
  - cva를 활용한 variant 및 size 버튼 스타일 유형 정의
- 커스텀 Carousel 컴포넌트 구현
  - 컴파운드 컴포넌트 패턴 기반 (ItemList, Item, Navigation 등)
  - 무한 스크롤 지원 (clone slide 활용)
  - default / centered / peek 모드 지원
  - 드래그 기반 슬라이드 이동
  - 반응형 슬라이드 개수 설정
  - active 슬라이드에 스타일 적용 가능
- 커스텀 Select 컴포넌트 구현
  - 컴파운드 컴포넌트 패턴 기반 (Trigger, Content, Item 등)
  - 제어/비제어 모드 지원
  - placement 기반 위치 지정
  - 포털 렌더링 및 외부 클릭 감지 처리
- 웰컴/장바구니/성공 모달 구현
  - 각 모달에 필요한 locale 텍스트 객체 분리 및 관리
  - 웰컴 모달에  react-hook-form을 활용하여 닉네임 유효성 검증 로직 추가
  - 웰컴 모달에 닉네임 저장 API 연동 및 성공 시 로그인 처리
  
## 체크리스트

- [x] 빌드에 성공했나요?
- [x] 코드 컨벤션을 잘 지켰나요? (예: `cmd + opt + L`)

## 테스트 결과
<p align="center">
  <img src="https://github.com/user-attachments/assets/e55b5d82-b836-4ea3-9b5d-151f1b18d5c3" width="48%" />
  <img src="https://github.com/user-attachments/assets/e64382f5-6fe0-4345-a8e3-88bb0e8fae46" width="48%" />
</p>

메인 페이지에서 각종 컴포넌트들을 확인 가능합니다

### 로그인 관련 오류 보고 및 미팅 사항 정리
로그인 후 웰컴 모달에서 닉네임을 작성하면 로그인 → 로그아웃 버튼으로 변경됨
→ 하지만 새로고침하면 store에 저장된 로그인 상태가 초기화되어
토큰이 있어도 로그인 버튼이 노출됨.
→ store를 persist사용하면 간편하긴 하지만 로컬스토리지에 상태여부를 저장하는 건 위험도가 있을 것으로 예상
```/**
 * db에서 유저정보를 가져와서 nickname 여부를 확인 후 login상태를 업데이트하기 위한 커스텀 훅입니다.
 * @returns userData: 유저정보 isLoading: 로딩 상태
 *
 */

export const useAuthStateManager = () => {
  const { data: userData, isLoading } = useGetUserInfo();
  const { setLogined } = useLoginedStore();

  useEffect(() => {
    if (isLoading) return;

    if (userData) {
      setLogined(userData.nickname !== null);
    } else {
      setLogined(false);
    }
  }, [userData, isLoading, setLogined]);

  return { userData, isLoading };
};
```

현재 커스텀 훅을 만든 상태라 이 함수를 Layout 컴포넌트에서 실행하는게 어떨 지 검토 중입니다.

minseung-gang committed:
Feat: Button 공통컴포넌트 구현

minseung-gang committed:
feat: cva를 사용해 variant/size 스타일 분기 처리
 버튼의 사이즈 옵션 (XL, L, M, S, XS) 적용
- variant 옵션에 border/green/orange 등 추가
- BaseButton과 StyledButton 분리 유지

minseung-gang committed:
fix: openModal 함수 매개변수 구조 분해 할당으로 수정

minseung-gang committed:
refactor: button 컴포넌트에 ref 전달 기능 추가

minseung-gang committed:
feat: Select 컴포넌트를 합성컴포넌트 패턴으로 구현

minseung-gang committed:
refactor: 캐러셀 컴포넌트 구조 개선 및 무한 스크롤/중앙 정렬 기능 통합

- CarouselContext에 다양한 상태 및 옵션 추가 (variant, isInfinite, actualIndex 등)
- variant에 따라 'default', 'centered', 'peek' 모드 지원
- visibleSlides props로 반응형 슬라이드 수 조절 가능
- 무한스크롤을 위한 cloneCount, actualIndex, adjustInfinitePosition 로직 추가
- 드래그 종료 시 이동 슬라이드 수 계산 개선 (slideWidth + gap 기준)
- 슬라이드 크기 및 위치 계산 함수 정리 및 통합 (calculateScrollX 등)

minseung-gang committed:
feat: 메인 배너 관련 이미지 추가

minseung-gang committed:
design: M, S, XS 사이즈 버튼에 w-fit 스타일 적용

minseung-gang committed:
refactor: Navigation 컴포넌트에서 className prop 제거 및 구조 단순화

minseung-gang committed:
refactor: Item 컴포넌트에 slideWidth 기반 inline width 적용 및 스타일 개선

minseung-gang committed:
refactor: ItemList 컴포넌트 구조 개선

- 무한 캐러셀을 위한 클론 생성 로직 추가
- 클래스네임 기반에서 인라인 스타일로 전환 (gap, transform)
- 원본/클론 구분을 위한 클래스 네임 부여

minseung-gang committed:
chore: classNames 유틸 함수 default import 적용

minseung-gang committed:
feat: Content 컴포넌트에 right 위치값 적용

minseung-gang committed:
remove: 디자인 시안 수정으로 사용하지 않는 파일 제거

minseung-gang committed:
design: Modal.Close 컴포넌트에 cursor-pointer 스타일 추가

minseung-gang committed:
feat: Select.Trigger 기능 확장
- icon 요소 외부에서 제어 가능하도록 추가
- icon 위치(left, right) 설정 기능 추가
- icon 숨김 기능 추가
- suffix(단위) 외부 제어 기능 추가

minseung-gang committed:
refactor: 불필요한 categoryBlock 타입 삭제하고 Category 타입만 사용하도록 정리

minseung-gang committed:
refactor: addToCart 타입 수정 및 welcomeStep 모달 타입 추가

minseung-gang committed:
feat: 상,하,좌,우 props를 받아 Trigger의 위치에서 팝업될 수 있도록 기능 추가

minseung-gang committed:
feat/refactor: Product 관련 타입 개선

- 기존 Product를 ProductItem으로 변경
- Product 타입에 hasNext, lastProductId, products 배열 추가

minseung-gang committed:
design: Select.Trigger컴포넌트에 transition 스타일 제거

minseung-gang committed:
feat: class-variance-authority 와 react-hook-form 라이브러리 추가

minseung-gang committed:
feat: MODAL_COMPONENTS에 welcomeStep모달 추가

minseung-gang committed:
refactor(carousel): cloneElement에 명확한 props 타입 적용

- any 타입 제거
- element에 className?: string 타입 확장
- CloneProps 인터페이스로 cloneElement 호출 시 타입 안정성 강화

minseung-gang committed:
fix: Item.List의 클론 Element의 key를 고유한 값으로 수정

minseung-gang committed:
Merge branch 'dev' of https://github.com/happy-pill/happyPill_Front into HP-149/ms

minseung-gang committed:
remove: 필요없는 파일 제거

minseung-gang committed:
design: variant XL size rounded 스타일 full로 수정

minseung-gang committed:
refactor: Navigation 컴포넌트 개선
- className props 추가로 외부에서 스타일 제어 가능
- 아이콘 교체 및 스타일 수정

minseung-gang committed:
refactor: Item 컴포넌트 className을 자식 요소로 이동

minseung-gang committed:
feat: Carousel에 active prop 추가하여 선택된 슬라이드 UI 제어

minseung-gang committed:
feat: ItemList 내 슬라이드에 active 상태에 따라 opacity 스타일 적용

minseung-gang committed:
feat: 장바구니 모달에 locale 적용 및 구독 옵션 유효성 검사 추가

minseung-gang committed:
feat: Modal.Close 컴포넌트에 onClick 핸들러 props 추가

minseung-gang committed:
refactor: 장바구니 담기 완료 모달 디자인 수정으로 인한 변경

minseung-gang committed:
feat: 메인페이지 웰컴모달 컴포넌트 구현
-locale 다국어 적용 완료 (CREATE_WELCOME_STEPS 기반)
- react-hook-form으로 유효성 검사 처리
- 닉네임 등록 API 연결 및 성공/실패 핸들링
- 닫기 버튼 누를 경우 로그아웃 및 모달 닫기 처리

minseung-gang committed:
feat: 메인 및 프로모션 배너, 모달 텍스트 등 locale 상수 객체 관리 추가

minseung-gang committed:
refactor: addToCart 모달,  cartAddSuccess 모달 props 수정

minseung-gang committed:
fix: tailwind-merge 확장하여 커스텀 폰트 사이즈 클래스 충돌 해결

minseung-gang committed:
rename: useLanguage > useLocale 파일명 수정

minseung-gang committed:
refactor: locale prop 제거하고 useLocale 훅으로 로케일 관리 통일

minseung-gang committed:
refactor: 언어 상태 관리 훅을 useLanguage에서 useLocale로 변경하여 명칭 통일

minseung-gang committed:
refactor:addToCart 모달, cartAddSUccess모달에서 locale props 제거

minseung-gang committed:
refactor: 모달 컴포넌트에서 locale props 제거하고 useLocale 훅으로 로케일 직접 조회하도록 변경

jieun419 commented on src/components/carousel/ItemList.tsx:
```diff
@@ -1,37 +1,111 @@
+
+    return [...frontClones, ...originalSlides, ...backClones];
+  };
+
```
frontClones, backClones, originalSlides 내부 로직이 반복 되는 것 같습니다.
반복되는 로직을 공통 함수로 빼고 관리하면 가독성이나 수정할 때도 좋을 듯 싶은데 어떠신가요?

jieun419 commented on src/components/modal/AddToCartModal.tsx:
```diff
@@ -1,66 +1,113 @@
+type CartProductProps = {
+  locale: 'ko' | 'en';
+  product: Omit<ProductItem, 'company' | 'categoryId'>;
+};
```
- 저희 컨벤션이 `interface`로 알고 있어서 CartProductProps를 `type`으로 선언하신 이유가 있을까요?
- locale type은 전역으로 타입을 생성 후 관리하는 거 어떠신가요? 다른 곳에서도 사용할 것 같아서요!

jieun419 commented on src/components/modal/WelcomeStepModal.tsx:
```diff
@@ -0,0 +1,133 @@
+import { logoutAndRedirect } from '@/utils/auth/logoutAndRedirect';
+import cn from '@/utils/classNames';
+
+interface Inputs {
```
Inputs는 포괄적 의미라 
form field의미로 `InputFields` 또는 
nickname만 사용한다면 `NicknamInput` 등의 
구체적인 명칭으로 변경하면 직관적일 듯 싶은데 어떠신가요?

jieun419 commented on src/components/modal/WelcomeStepModal.tsx:
```diff
@@ -0,0 +1,133 @@
+  const { closeModal } = useModal();
+  const { setLogined } = useLoginedStore();
+  const stepMessages = CREATE_WELCOME_STEPS;
+  const currentMessage = stepMessages[currentStep];
```
CREATE_WELCOME_STEPS를 stepMessages에 변수로 한번 더 담으신 이유가 가독성이나 이후 확장성 고려 때문 일까요?

별다른 이유가 없다면 `CREATE_WELCOME_STEPS[currentStep]` 으로 가는 것도 좋을 것 같아서요!

minseung-gang replied:
네 지은님 말씀처럼 가독성을 위해 한 번 더 변수로 담았던 건데, 저도 이 부분이 살짝 고민되긴 했었어요ㅎㅎ 
지금 말씀해주신 방향대로 바로 사용하는 것도 더 깔끔할 것 같아서 반영해보겠습니다.

jieun419 commented on src/components/modal/WelcomeStepModal.tsx:
```diff
@@ -0,0 +1,133 @@
+  return (
+    <Modal>
+      <Modal.Content>
+        <Modal.Close className='absolute top-1 right-1' onClick={() => handleClose()}>
```
handleClose에 전달해야하는 데이터가 없다면
`onClick={() => handleClose()}` -> `onClick={handleClose}`로 가도  될 것 같아요!

jieun419 commented on src/components/select/Trigger.tsx:
```diff
@@ -0,0 +1,61 @@
+  className?: string;
+  icon?: React.ReactNode; //아이콘(컴포넌트)
+  iconPosition?: 'left' | 'right'; // 아이콘의 위치
+  hideIcon?: boolean; // 아이콘은 없앨 지
```
타입이 boolean으로 `isHideIcon`이 의미나 가독성 측면에서 알맞을 것 같아요!

jieun419 commented on src/components/select/Trigger.tsx:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    // 기본 아이콘만 rotate 애니메이션 적용
+    return <IoIosArrowDown className={cn(isOpen && 'rotate-180')} />;
```
cn 유틸 함수는 여러 클래스들이 들어갈 때 중복 제거하는 걸로 알고 있습니다.
IoIosArrowDown의 클래스는 isOpen 여부에 따른 단일 클래스로 
클래스가 하나만 적용 되는거라면

`<IoIosArrowDown className={isOpen ? 'rotate-180' : ''} /`> 삼항 연산자등을 사용해도 좋을 것 같아요!

minseung-gang replied:
네 지은님,말씀해주신 대로 지금은 삼항이 더 간결할 수 있어요!
다만 이후 클래스가 추가될 걸 고려해 `cn()` 패턴으로 작성해뒀습니다 :) 
둘 다 맞는 방식이라 팀 기준에 맞춰가면 될 것 같아요!

jieun419 commented on src/components/select/Trigger.tsx:
```diff
@@ -0,0 +1,61 @@
+      )}
+    >
+      {iconPosition === 'left' && iconElement && renderIcon()}
+      <span>{selectedValue != null ? displayValue : placeholder}</span>
```
`!=` 보단 사용 권장, 엄격한 비교를 위해  `!==`  사용하는게 좋을 것 같아요!
혹시 `!=` 를 사용한 이유가 있을까요?

minseung-gang replied:
빠르게 코드 작성하면서 디테일한 부분을 놓쳤던 것 같습니다. 
지은님께서 짚어주신 부분이 정확해요! 피드백 감사합니다!

jieun419 commented on src/hooks/useLocale.ts:
아까 코멘트와 동일합니다.

locale 타입 'ko' | 'en'을 전역으로 사용할 수 있도록 관리하면 좋을 것 같아요!

jieun419 reviewed (commented):
고생하셨습니다!
- `useAuthStateManager`를 Layout에 배치하는 것 좋다 생각합니다!

minseung-gang committed:
refactor: Carousel 코드 리팩토링
- 함수에 useCallback함수를 사용하여 리렌더링시 불필요한 함수 재생성 방지
- visibleSlides에 null을 허용하여 유연한 자동 슬라이드 갯수 설정 지원
- 컨벤션 규칙에 맞게 useEffect 함수를 코드 맨 마지막에 배치

minseung-gang committed:
refactor: 중복된 클론 생성 로직을 재사용 가능한 함수로 분리
- buttonClones 함수 추출로 코드 중복 제거
- 캐러셀 클론 생성 로직의 유지보수성 향상

minseung-gang committed:
refactor: locale 타입 공통 분리 및 AddToCartModal에 적용

minseung-gang committed:
feat: locale 타입 분리 및 공통 관리

minseung-gang committed:
refactor: 닉네임 입력 전용 useForm 타입 NicknameInput으로 변경 및 적용

minseung-gang committed:
refactor: CREATE_WELCOME_STEPS 직접 접근으로 불필요한 변수 제거

minseung-gang committed:
refactor: 불필요한 래퍼 함수 제거

minseung-gang committed:
refactor: Trigger 컴포넌트 리팩토링
- hideIcon prop을 isHideIcon으로 변경
- cn 함수 대신 삼항 연산자로 rotate-180 스타일 적용

minseung-gang committed:
refactor: 전역 LocaleType을 사용하도록 useLocale 수정

minseung-gang committed:
refactor: LocaleProps를 LocaleType으로 이름 변경

minseung-gang committed:
refactor: LocaleProps를 LocaleType으로 변경에 따른 타입 수정

minseung-gang committed:
refactor: != 를 !== 로 변경하여 엄격한 비교 적용

jieun419 reviewed (approved):
고생하셨슴다!
